view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java

```diff
@@ -1,58 +1,108 @@
+import static org.hypertrace.core.viewgenerator.service.ViewGeneratorConstants.KAFKA_STREAMS_CONFIG_KEY;
+import static org.hypertrace.core.viewgenerator.service.ViewGeneratorConstants.OUTPUT_TOPIC_CONFIG_KEY;
+import static org.hypertrace.core.viewgenerator.service.ViewGeneratorConstants.VIEW_GENERATOR_CLASS_CONFIG_KEY;
+
+import com.typesafe.config.Config;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.kafka.common.serialization.Serdes;
+import org.apache.kafka.streams.StreamsBuilder;
+import org.apache.kafka.streams.kstream.Consumed;
+import org.apache.kafka.streams.kstream.KStream;
+import org.apache.kafka.streams.kstream.Produced;
+import org.hypertrace.core.kafkastreams.framework.KafkaStreamsApp;
 import org.hypertrace.core.serviceframework.config.ConfigClient;
+import org.hypertrace.core.serviceframework.config.ConfigUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-/**
- * ViewGeneratorLauncher reads enriched structured traces and converts them to Pinot records using specified Avro
- * schema specified in the target view generator class, which is specified in the config.
- */
-public class ViewGeneratorLauncher extends PlatformService {
-
-  private static Logger LOGGER = LoggerFactory.getLogger(ViewGeneratorLauncher.class);
+public class ViewGeneratorLauncher extends KafkaStreamsApp {
 
-  private ViewGenerationJob job;
-  private StreamExecutionEnvironment environment;
+  private static final Logger logger = LoggerFactory.getLogger(ViewGeneratorLauncher.class);
+  private static final String DEFAULT_VIEW_GEN_JOB_CONFIG_KEY = "view-gen";
+  private String viewGenName;
 
   public ViewGeneratorLauncher(ConfigClient configClient) {
     super(configClient);
   }
 
+  public String getViewGenName() {
+    return viewGenName;
+  }
+
+  public void setViewGenName(String viewGenName) {
+    this.viewGenName = viewGenName;
+  }
+
   @Override
-  protected void doInit() {
+  public StreamsBuilder buildTopology(Map<String, Object> properties, StreamsBuilder streamsBuilder,
+      Map<String, KStream<?, ?>> inputStreams) {
+
+    Config jobConfig = getJobConfig(properties);
+    String viewGeneratorClassName = jobConfig.getString(VIEW_GENERATOR_CLASS_CONFIG_KEY);
+
+    InputToViewMapper viewMapper;
     try {
-      environment = FlinkEnvironmentFactory.createExecutionEnv(getAppConfig());
-      job = new ViewGenerationJob(environment, getAppConfig());
-      job.init();
+      viewMapper = new InputToViewMapper(viewGeneratorClassName);
     } catch (Exception e) {
-      LOGGER.error("Failed to initialize ViewGenerationJob.", e);
       throw new RuntimeException(e);
     }
+
+    String inputTopic = jobConfig.getString(INPUT_TOPIC_CONFIG_KEY);
+    String outputTopic = jobConfig.getString(OUTPUT_TOPIC_CONFIG_KEY);
+
+    KStream<?, ?> inputStream = inputStreams.get(inputTopic);
```

Comment:
Can we avoid wildcard generics if we can?

Reply:
we would like to merge our ingester pipeline as one stream in hypertrace-ingester. Here, we will have different input streams types with RawSpan, StructuredTrace, so I am not very sure that we can base templatize or have the interface with Object.  @surajpuvvada  what do you think? The change will need an interface change of buildTopology, so, I would suggest doing as follow up to finish with the first e2e pipeline.

Reply:
yes the variables could be of multiple types @tim-mwangi do you have any suggestions ?